### DISCUSSION TITLE
Typing custom form components

### DISCUSSION THREAD

**Discussion by RWOverdijk** [TanStack/form#922] [Q&A] [answered]
I am trying to create my own components since tanstack form is great, but pretty verbose. I'm using it with react native.

I am trying to get one of the following uses.

**Preferred:**

```tsx
<FormInput
  name="name"
  label="Business name"
  placeholder="Hat rentals"
  form={form}
  validators={{
    onSubmit: v.pipe(v.string(), v.minLength(3)),
  }}
/>
```

**Alternatively (assuming field.name equals name):**

```tsx
<form.Field name="name" validators={{ onSubmit: v.pipe(v.string(), v.minLength(3)) }}>
  {(field) => (
    <FormInput field={field} label="Business name" placeholder="Festi hats rental" />
  )}
</form.Field>
```

My biggest issue is with typing. I can't figure out how to properly type form or field. I've looked through the docs and other issues but the closest I could find was https://github.com/TanStack/form/discussions/636 and that didn't help me.

Any help or guidance would be appreciated.

**Comment by Balastrong** [comment, marked as answer]
You may find some hints in #825 and the `ReactFormApi` type.

We'd like to come out with a supported/recommended way of creating custom components/wrappers at some point, feedback is welcome! :)

**Reply by RWOverdijk** [reply]
Ah, cool. I see that adds type changes that I do not have yet as well.

It's hard to give feedback at this point then. Also doubt my opinion would be shared by many, since I prefer dx over performance when it comes to forms haha. I'll patiently wait :)

ETA: By the power of `any` I can do this:

```tsx
<form.Field
  name="name"
  validators={{ onSubmit: v.pipe(v.string(), v.minLength(2)) }}
>
  {(field) => (
    <FormInput
      label="Business name"
      placeholder="Festi hats rental"
      field={field}
    />
  )}
</form.Field>
```

Which is good enough for me. I don't really care about the typing inside of the FormInput component enough to worry about this for now. 😄

**Reply by Balastrong** [reply]
That's great!

You should probably be able to type `field` as well with something along the lines of `ReactFormApi<FormData, ZodValidator>["Field"]`

**Reply by Mnigos** [reply]
> That's great!
> 
> You should probably be able to type `field` as well with something along the lines of `ReactFormApi<FormData, ZodValidator>["Field"]`

Not really, you have to pass 10 type arguments
```
Generic type 'ReactFormApi<TFormData, TOnMount, TOnChange, TOnChangeAsync, TOnBlur, TOnBlurAsync, TOnSubmit, TOnSubmitAsync, TOnServer, TSubmitMeta>' requires 10 type argument(s).
```

**Reply by Balastrong** [reply]
Yeah that answer is almost one year old, we weren't 1.0 yet and many things have changed... and improved!

Rather than playing with the types you can achieve much more with [Form Composition](https://tanstack.com/form/latest/docs/framework/react/guides/form-composition) now.

**Reply by Mnigos** [reply]
Thanks for the answer! I'm aware of the form composition, but currently I'm using [tanstack form with shadcnui](https://fatahchan.github.io/shadcn-tanstack-form/). And I was thinking if I can achieve typesafety without needing to redeclare the form component. 

I was thinking about a way to somehow infer the type from hook and then passing field name as type argument so fe. 

```ts
	const form = useAppForm({
		defaultValues: {
			content: initialData?.content ?? '',
			projectId: initialData?.projectId,
			priority: initialData?.priority ?? 1,
			labels: initialData?.labels ?? [],
			dueDate: initialData?.dueDate,
			description: initialData?.description ?? '',
		},
	})
	
	type TaskFormField<TFieldName extends string> = (inferFormApi<typeof form>)<TFieldName>
```
Where the `inferFormApi` would be new type util for inferring FormApi from created form hook.

I achieved something like that with current library implementation - everything was properly typed except it had a union of all fields. So the `field.state.value` was a union of fe. `string` and `number`.

I'm not sure If I explained it properly, but what do you think about it? I can provide you more detailed example, but I would like to know your thoughts about this first. Could it be added or this would be considered as anti pattern?

**Comment by uzakovnikita** [comment]
```
import { useAppForm, ZodFieldInfo } from "@shared";

import { z } from "zod";

export const clientNameSchema = z
  .string({
    required_error: "required",
    invalid_type_error: "not valid format",
  })
  .min(3, "name should have length more than 3")
  .max(10, "name should have length less than 11");

export const NameField = ({
  form,
  name,
}: {
  name: string;
  form: ReturnType<typeof useAppForm>;
}) => {
  return (
    <div>
      <form.AppField
        name={name}
        validators={{
          onBlur: clientNameSchema,
        }}
      >
        {(field) => (
          <>
            <field.TextField label="Ваше имя" />
            <ZodFieldInfo field={field} />
          </>
        )}
      </form.AppField>
    </div>
  );
};

```

Hello! I'm trying to make custom components that will contain validation logic inside themselves, but I have a problem with typing the "form" prop. I didn't find a suitable solution in the documentation, please tell me if there is an API that solves my problem?

**Comment by oliverstreissi** [comment]
Is there an update on this? Tried to achieve the same as @uzakovnikita but failed to solve it.

**Comment by Frost7994** [comment]
I'm looking for an update on this too like @oliverstreissi  and @uzakovnikita

**Comment by mamlzy** [comment]
How to achieve this https://github.com/TanStack/form/discussions/922#discussioncomment-13044725? i think this is very important @Balastrong @LeCarbonator @crutchcorn. Sorry for tagging.

**Reply by LeCarbonator** [reply]
Looks like a case for `withFieldGroup`. You can combine fields with logic and dependent fields.

**Reply by mamlzy** [reply]
> Looks like a case for `withFieldGroup`. You can combine fields with logic and dependent fields.

Yeah, I’m using `withForm` here’s an example video: https://youtu.be/H2T21r5wu3g?t=7586
@uzakovnikita @oliverstreissi @Frost7994

**Reply by Frost7994** [reply]
@mamlzy @LeCarbonator this isn't a solution as I must pass the same formOpts to each sub component .. looking for a way to create a component that accepts a form properly that can have any formOpts passed so that it can fit into multiple forms in my app rather than be tied in with one specific form, so need the generic form return type or something similar

**Reply by LeCarbonator** [reply]
Yes, `withFieldGroup` is not tied to one form specifically. You're talking about `withForm` which is indeed hooked to one form type. [Docs link](https://tanstack.com/form/latest/docs/framework/react/guides/form-composition#reusing-groups-of-fields-in-multiple-forms)

**Reply by Frost7994** [reply]
I know but with field group is tying component logic such as password and confirm password. I'm looking to create a component for example called generic input that takes a form prop to be used for form logic and other props that are used in the component such as label etc. The idea being a multi step form, currently the implementation has to be that each step is it's own form. I want each form to validate that steps data and push to the next step. Each step has the exact same markup with different data i.e an email updates field and an email communications field. Therefore I need to create a reusable component to keep the whole process DRY but I can't genetically type the form prop. Does this make more sense without a code example?

**Reply by LeCarbonator** [reply]
A code example would be great. I don't want to assume too much about the end goal and derail the conversation.

**Reply by Frost7994** [reply]
Yeah perfect. This is a sample example.

```
"use client";

import z from "zod";

import { useAppForm } from "@/hooks/forms";

type Props = {
  section: Section;
  form: // CAN'T GET THE TYPE FOR THIS
};

const GenericComponent = ({section, form}:Props) => {
  return (
    <FormSection>
      <FormSectionHeader>
        <FormSectionTitle>{section.title}</FormSectionTitle>
        <FormSectionDescription>{section.description}</FormSectionDescription>
      </FormSectionHeader>
      <FormSectionContent>
        {section.settings.map((setting) => {
          return (
            <form.AppField key={setting.id} name={setting.schemaName}>
              {(field) => <field.SettingsSwitch setting={setting} />}
            </form.AppField>
          );
        })}
      </FormSectionContent>
    </FormSection>
  );
};

const StepOne = () => {
  const form = useAppForm({
    defaultValues: {
      emailCommunication: true,
      emailUpdates: true,
    },
    validators: {
      onSubmit: z.object({
        emailCommunication: z.boolean(),
        emailUpdates: z.boolean(),
      }),
    },
    onSubmit: ({ value }) => {
      console.log(value);

      // SAVE VALUES TO STORE AND PUSH TO NEXT STEP
    },
  });

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        form.handleSubmit();
      }}
    >
      <GenericComponent form={form} data={data} />
    </form>
  );
};

const StepTwo = () => {
  const form = useAppForm({
    defaultValues: {
      lightMode: false,
      reducedAnimation: true,
    },
    validators: {
      onSubmit: z.object({
        lightMode: z.boolean(),
        reducedAnimation: z.boolean(),
      }),
    },
    onSubmit: ({ value }) => {
      console.log(value);

      // SAVE VALUES TO STORE AND PUSH TO NEXT STEP
    },
  });

  return (
    <form
      onSubmit={(e) => {
        e.preventDefault();
        form.handleSubmit();
      }}
    >
      <GenericComponent form={form} data={data} />
    </form>
  );
};

```

**Reply by kein-1** [reply]
I'm, also still having trouble typing the form and the AppField +  subField API if i want to pass as props; 

Not really sure how to solve it other than using the withForm as mentioned in form composition docs but I've really only used that to break up larger forms (i.e i have a big form but show each half on one screen)

**Reply by ihsanhere** [reply]
Maybe we can do something like this since `formOptions` helper already provides a typesafe way. In most examples it shows we can share the `myFormOptions` object.

```typescript
const myFormOptions = formOptions({
  defaultValues: {
    name: '',
    location: '',
    age: 0,
  }
})
const MyFormType = InferFormType<typeof myFormOptions>
```

**Comment by jaskang** [comment]
I have a working example of this implementation here that you might find helpful: https://github.com/TanStack/form/discussions/2031

**Comment by Desalutar20** [comment]
Any news ? How we can import type FieldApi for example ?

**Reply by LeCarbonator** [reply]
In v2, we'll expose helper types to split up forms and fields. `FieldWithValue<T>` and `ReactFormType<TOptions>` for example.
